src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_IncrementalAnalyzer.cs

```diff
@@ -368,9 +376,15 @@ private IReadOnlyList<StateSet> GetStateSetsForFullSolutionAnalysis(IEnumerable<
             // If full analysis is off, remove state that is created from build.
             // this will make sure diagnostics from build (converted from build to live) will never be cleared
             // until next build.
-            if (GlobalOptions.GetBackgroundAnalysisScope(project.Language) != BackgroundAnalysisScope.FullSolution)
+            _ = GlobalOptions.IsFullSolutionAnalysisEnabled(project.Language, out var compilerFullSolutionAnalysisEnabled, out var analyzersFullSolutionAnalysisEnabled);
+            if (!compilerFullSolutionAnalysisEnabled)
+            {
+                stateSets = stateSets.Where(s => !s.Analyzer.IsCompilerAnalyzer() || !s.FromBuild(project.Id));
+            }
+
+            if (!analyzersFullSolutionAnalysisEnabled)
             {
-                stateSets = stateSets.Where(s => !s.FromBuild(project.Id));
+                stateSets = stateSets.Where(s => s.Analyzer.IsCompilerAnalyzer() || !s.FromBuild(project.Id));
             }
```

Comment:
this code is subtle different from teh last check i say.  hwere, both of teh out values could be null, and we'll run both block bodies.  but in teh first check i saw (here: https://github.com/dotnet/roslyn/pull/60808/files#diff-341e6fd63c9ee5257e195befbb7ba9151217bad7279bd1a02c415d9c77b06e2eR191), we only run one block.  Intentional?

Reply:
Yes, the latest version of this code is the one we want. The prior one was indeed wrong and led to test failures. We want to filter out all regular analyzer statesets which are not from build, without touching the compiler analyzer stateset.